### DISCUSSION TITLE
url_for and Query params

### DISCUSSION THREAD

**Discussion by capkz** [fastapi/fastapi#5967] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.responses import HTMLResponse
from pydantic import IPvAnyNetwork
from fastapi import Request, Query

app = FastAPI()

@router.get("/{datacenter}/palo-alto/health-check/", response_class=HTMLResponse)
async def health_check(request: Request, datacenter, ci_ip: List[IPvAnyNetwork] = Query()):
return True
```

### Description

Trying to get the URL for this endpoint within Jinja, by using `{{ url_for('health_check', datacenter=datacenter, ci_ip=['10.214.234.146', '10.214.234.145']) }}`

However I get an error saying: `starlette.routing.NoMatchFound: No route exists for name "health_check" and params "datacenter, ci_ip".`

### Operating System

macOS

### FastAPI Version

0.88.0

### Python Version

3.10

**Comment by obahamonde** [comment]
Hi @capkz IPv4AnyNetwork comes from pydantic module, which makes validations on the request Body, not the query, you should pass the cidr block as part of the request body to allow pydantic to validate it, also when called inside Jinja context, IPv4AnyNetwork calls either the str of repr method that might contain extrametadata so you should cast to str before serializing to jinjaenv.

**Reply by capkz** [reply]
Hey @obahamonde thanks for your reply, I removed the code to not include IPv4AnyNetwork in it as a test, just a list of query params like so: `async def health_check(request: Request, datacenter, ci_ip: list = Query()):`
but still, `url_for()` does not seem to work, is `url_for()` in Jinja context not capable of handling query parameters?

**Reply by obahamonde** [reply]
Hi @capkz actually url_for is a Flask feature that it's not present in fastapi because it doesn't rely on named endpoint functions atm.

**Comment by nymous** [comment]
`url_for` in Starlette doesn't yet support query parameters. You can follow the issues tiangolo/fastapi#323 and encode/starlette#560, and they also offer some workarounds (the FastAPI issue has a more advanced Jinja helper with URL validation).

On a side note, you *can* use an `IPv4AnyNetwork` type annotation for a query parameter, and it will be validated by Pydantic!

**Reply by capkz** [reply]
Thanks for the provided links, looking through them, sad to see Starlette doesn't support query parameters yet.

**Comment by caeser1996** [comment]
Can you please try @capkz 
```
@app.get("/{datacenter}/palo-alto/health-check/", response_class=HTMLResponse, name="health_check")
async def health_check(request: Request, datacenter, ci_ip: List[IPvAnyNetwork] = Query()):
    return True
```
Now you should be able to generate the URL for the endpoint using url_for('health_check', datacenter=datacenter, ci_ip=['10.214.234.146', '10.214.234.145']) in Jinja.

**Reply by capkz** [reply]
Thank you for your response, unfortunately this did not work for this case as well.

**Comment by Kludex** [comment, marked as answer]
https://github.com/encode/starlette/pull/1385
